bot/seasons/evergreen/space.py

```diff
@@ -0,0 +1,217 @@
+import logging
+import random
+from datetime import datetime
+from typing import Any, Dict, Optional, Union
+from urllib.parse import urlencode
+
+from discord import Embed
+from discord.ext import tasks
+from discord.ext.commands import BadArgument, Cog, Context, Converter, group
+
+from bot.bot import SeasonalBot
+from bot.constants import Tokens
+
+logger = logging.getLogger(__name__)
+
+NASA_BASE_URL = "https://api.nasa.gov"
+NASA_IMAGES_BASE_URL = "https://images-api.nasa.gov"
+NASA_EPIC_BASE_URL = "https://epic.gsfc.nasa.gov"
+
+APOD_DEFAULT_PARAMS = {
+    "api_key": Tokens.nasa
+}
+
+
+class DateConverter(Converter):
+    """Parse SOL or earth date (in format YYYY-MM-DD) into `int` or `datetime`. When invalid input, raise error."""
+
+    async def convert(self, ctx: Context, argument: str) -> Union[int, datetime]:
+        """Parse date (SOL or earth) into `datetime` or `int`. When invalid value, raise error."""
+        if argument.isdigit():
+            return int(argument)
+        try:
+            date = datetime.strptime(argument, "%Y-%m-%d")
+        except ValueError:
+            raise BadArgument(f"Can't convert `{argument}` to `datetime` in format `YYYY-MM-DD` or `int` in SOL.")
+        return date
+
+
+class Space(Cog):
+    """Space Cog contains commands, that show images, facts or other information about space."""
+
+    def __init__(self, bot: SeasonalBot):
+        self.bot = bot
+        self.http_session = bot.http_session
+
+        self.rovers = {}
+        self.get_rovers.start()
+
+    @tasks.loop(hours=24)
+    async def get_rovers(self) -> None:
+        """Get listing of rovers from NASA API and info about their start and end dates."""
+        data = await self.fetch_from_nasa("mars-photos/api/v1/rovers", params={"api_key": Tokens.nasa})
+
+        for rover in data["rovers"]:
+            self.rovers[rover["name"].lower()] = {
+                "min_date": rover["landing_date"],
+                "max_date": rover["max_date"]
+            }
+
+    @group(name="space", invoke_without_command=True)
+    async def space(self, ctx: Context) -> None:
+        """Head command that contains commands about space."""
+        await ctx.send_help("space")
+
+    @space.command(name="apod")
+    async def apod(self, ctx: Context, date: Optional[str] = None) -> None:
+        """
+        Get Astronomy Picture of Day from NASA API. Date is optional parameter, what formatting is YYYY-MM-DD.
+
+        If date is not specified, this will get today APOD.
+        """
+        params = APOD_DEFAULT_PARAMS.copy()
```

Comment:
Instead of copying this constant in one subcommand, recreating it entirely in another subcommand, etc. , how about we have the `fetch_from_nasa` method create the base parameters and add in any additional parameters required? This way the behavior is consistent between all of the subcommands.